src/ReducedModelControllers/src/CentroidalMPC.cpp

```diff
@@ -0,0 +1,1341 @@
+            {
+                using namespace casadi;
+                ddcom += contact.isEnabled / mass * corner.force;
+                angularMomentumDerivative
+                    += contact.isEnabled
+                       * MX::cross(MX::mtimes(MX::reshape(contact.orientation, 3, 3),
+                                              corner.position)
+                                       + contact.position - com,
+                                   corner.force);
+
+                input.push_back(corner.force);
+            }
+        }
+
+        const double dT = chronoToSeconds(this->optiSettings.samplingTime);
+
+        std::vector<std::string> outputName{"com", "dcom", "angular_momentum"};
+        std::vector<casadi::MX> rhs{com + dcom * dT,
+                                    dcom + ddcom * dT,
+                                    angularMomentum + angularMomentumDerivative * dT};
+
+        for (const auto& [key, contact] : casadiContacts)
+        {
+            rhs.push_back(contact.position + (1 - contact.isEnabled) * contact.linearVelocity * dT);
+            outputName.push_back(key);
+        }
+
+        return casadi::Function("centroidal_dynamics",
+                                std::move(input),
+                                std::move(rhs),
+                                extractVariablesName(input),
+                                std::move(outputName));
+    }
+
+    casadi::Function contactPositionError()
+    {
+        casadi::MX contactPosition = casadi::MX::sym("contact_position", 3);
+        casadi::MX nominalContactPosition = casadi::MX::sym("nominal_contact_position", 3);
+        casadi::MX contactOrientation = casadi::MX::sym("contact_orientation", 3 * 3);
+
+        // the orientation is stored as a vectorized version of the matrix. We need to reshape it
+        casadi::MX rhs = casadi::MX::mtimes(casadi::MX::reshape(contactOrientation, 3, 3).T(),
+                                            contactPosition - nominalContactPosition);
+
+        return casadi::Function("contact_position_error",
+                                {contactPosition, nominalContactPosition, contactOrientation},
+                                {rhs},
+                                extractVariablesName({contactPosition, //
+                                                      nominalContactPosition,
+                                                      contactOrientation}),
+                                {"error"});
+    }
+
+    void resizeControllerInputs()
+    {
+        constexpr int vector3Size = 3;
+        const int stateHorizon = this->optiSettings.horizon + 1;
+
+        std::size_t vectorizedOptiInputsSize = 7 + (this->output.contacts.size() * 6);
+
+        if (this->optiSettings.isWarmStartEnabled)
+        {
+            // in this case we need to add the com, the angular momentum and the contact location
+            vectorizedOptiInputsSize += 2 + (this->output.contacts.size() * 1);
```

Comment:
because here I'm counting the vectors not the element of the vectors so:
- `1` is the com 
- `1` is the angular momentum
- and then for each contact I have its position (that's the reason why there is `* 1`)

Reply:
I am confused, does the same variables store the number of elements in the line above? https://github.com/ami-iit/bipedal-locomotion-framework/blob/af5760b6695136ee3845dff5939cedf4c22e8ec6/src/ReducedModelControllers/src/CentroidalMPC.cpp#L543

Reply:
So in case of no warmstart the variables are
- `7`: external force + external torque + com current + dcom current + angular momentum current + com reference + angular momentum reference
- `(this->output.contacts.size() * 6)`: for each contact we have current position + nominal position + orientation + is enabled + upper limit in position + lower limit in position

If the warmstart is enabled we need to add `2 + (this->output.contacts.size() * 1)`
- `1` is the com 
- `1` is the angular momentum
- and then for each contact I have its position (that's the reason why there is `* 1`)

Reply:
Ah, I see. I was getting confused on the actual content of the vectorized input. I would suggest just to have it commented somewhere